### ISSUE TITLE
Bug: [no-unused-vars] False positive on function parameter on types and interfaces

### ISSUE THREAD

**Issue by jmezzera** [typescript-eslint/typescript-eslint#7601]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=C4TwDgpgBAglC8UDeAoKUCGAuKAKARjgM7ABOAlgHYDmAlAgHxQBuA9uQCYoC%2BKKAxq0olMOOIlTpsUfIyiDhrADYQAdEtbUCtHgKFFlajVoy0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWXWZRAE0IDcBDaVDSfAMxcnDAF8Qug&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
type A = {
  a: (b: string) => void
}

const a: A = {
  a: b => console.log(b)
}
console.log(a)
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "no-unused-vars": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the type `(b: string) => void` to be valid

### Actual Result

ESLint is complaining about function parameter not being used.

**Comment by JoshuaKGoldberg** [comment]
> - [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

See the first heading under that link: https://typescript-eslint.io/linting/troubleshooting/#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code

**Comment by jmezzera** [comment]
🤦 

Thks!!
